util/sha256fixed/sha256.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package sha256fixed is like crypto/sha256,
+// but efficiently hashes fixed-width integers.
+package sha256fixed
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"hash"
+)
+
+var _ hash.Hash = (*Hash)(nil)
+
+// Hash is a hash.Hash for SHA-256,
+// but has efficient methods for hashing fixed-width integers.
+type Hash struct {
+	// The optimization is to maintain our own block and
+	// only call h.Write with entire blocks.
+	// This avoids double-copying of buffers within sha256.digest itself.
+	// However, it does mean that sha256.digest.x goes unused,
+	// which is a waste of 64B.
+
+	h  hash.Hash              // always *sha256.digest
+	x  [sha256.BlockSize]byte // equivalent to sha256.digest.x
+	nx int                    // equivalent to sha256.digest.nx
+}
+
+func New() *Hash {
```

Comment:
how much of the win is that we're using a concrete type and not going through interface indirect calls?

Reply:
The 40% win is purely from avoiding the double copy.
The benchmarks currently go through an interface to keep them fairer.
When we consider the fact that we will actually be calling concrete methods,
the performance gains are even better at 140%. 😄

Reply:
hell yeah